snapcraft/_options.py

```diff
@@ -193,18 +193,31 @@ def get_core_dynamic_linker(self):
         However if core is not installed None will be returned.
         """
         core_path = common.get_core_path()
-        core_dynamic_linker = self.__machine_info.get('core-dynamic-linker',
-                                                      'lib/ld-linux.so.2')
-
-        try:
-            dynamic_linker_resolved_path = os.readlink(
-                os.path.join(core_path, core_dynamic_linker))
-            dynamic_linker_path = os.path.join(
-                core_path, dynamic_linker_resolved_path.lstrip('/'))
-        except FileNotFoundError:
-            dynamic_linker_path = None
-
-        return dynamic_linker_path
+        dynamic_linker_path = os.path.join(
+            core_path,
+            self.__machine_info.get('core-dynamic-linker',
+                                    'lib/ld-linux.so.2'))
+
+        # We can't use os.path.realpath because any absolute symlinks
+        # have to be interpreted relative to core_path, not the real
+        # root.
+        seen_paths = set()
+        while True:
+            if dynamic_linker_path in seen_paths:
+                raise EnvironmentError(
```

Comment:
This won't do anything about the errno in `EnvironmentError`. Can we please use `snapcraft.internal.errors.SnapcraftEnvironmentError` instead?